### PR TITLE
Show functionality added to automatic DeepLink Controller - AIS-254

### PR DESCRIPTION
**Pull request by parthkalavadia**
Automatic DeepLink Controller will have three options now- push, present or show, just like any other viewcontroller.

### CODE REVIEW

**Comment by BranchMacMini** [comment]
Result of [Integration 1](https://stlt.herokuapp.com/v1/xcs_deeplink/qabot.stage.branch.io/b388335146db7e936b7215574da2f2b2/b388335146db7e936b7215574da2fa45)
---
*Duration*: 1 minute and 58 seconds
*Result*: **1 error, failing state: build-errors**

**Comment by BranchMacMini** [comment]
Result of [Integration 2](https://stlt.herokuapp.com/v1/xcs_deeplink/qabot.stage.branch.io/b388335146db7e936b7215574da2f2b2/b388335146db7e936b7215574da39e1a)
---
*Duration*: 1 minute and 18 seconds
*Result*: **1 error, failing state: build-errors**

**Comment by BranchMacMini** [comment]
Result of [Integration 3](https://stlt.herokuapp.com/v1/xcs_deeplink/qabot.stage.branch.io/b388335146db7e936b7215574da2f2b2/91a3adb793e487e2a3459e663292a354)
---
*Duration*: 4 minutes and 3 seconds
*Result*: **Perfect build!** All 179 tests passed. :+1:
*Test Coverage*: 56%

**Comment by E-B-Smith** [comment]
What is the lowest iOS OS version that this is compatible with?  Right now we support down to iOS 7.

**Comment by parthkalavadia** [comment]
Yes, It does supported iOS 7. I have put the neccessary checks for it: 
`if ([self.deepLinkPresentingController isKindOfClass:[UINavigationController class]]) {
                        if ([self.deepLinkPresentingController respondsToSelector:@selector(showViewController:sender:)]) {
                            [self.deepLinkPresentingController showViewController:branchSharingController sender:self];
                        }else{
                            [(UINavigationController*)self.deepLinkPresentingController pushViewController:branchSharingController animated:true];
                        }
                    }else{
                        [self presentSharingViewController:branchSharingController];
                }
`

File: Branch-SDK/Branch-SDK/Branch.h [hunk 1]
```diff
@@ -399,7 +400,14 @@ typedef NS_ENUM(NSUInteger, BranchCreditHistoryOrder) {
...
 
-- (void)registerDeepLinkController:(UIViewController <BranchDeepLinkingController> *)controller forKey:(NSString *)key;
+
+/**
+ Allow Branch to handle a view controller with options to push, present or show.
+ Note:
+ * If push option is used and the rootviewcontroller of window is not of type UINavigationViewController, than the sharing View controller would be presented automatically
+ * If UIViewController is pushed, than it wouldn't be popped on calling deepLinkingControllerCompleted delegate
+ */
+- (void)registerDeepLinkController:(UIViewController <BranchDeepLinkingController> *)controller forKey:(NSString *)key withOption:(BNCViewControllerOption)option;
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/Branch.h]: It's be important that you don't change the interface without warning. If a developer were to upgrade, their build would break. You need to make sure you continue to support the old method.

File: Branch-SDK/Branch-SDK/BranchDeepLinkModel.h [hunk 2]
```diff
@@ -0,0 +1,16 @@
+//
+//  BranchDeepLinkModel.h
+//  Foodie Woodie
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/BranchDeepLinkModel.h]: Can you remove this project name from the header?

File: Branch-SDK/Branch-SDK/BranchDeepLinkingController.h [hunk 3]
```diff
@@ -5,13 +5,21 @@
 //  Created by Graham Mueller on 6/18/15.
 //  Copyright (c) 2015 Branch Metrics. All rights reserved.
 //
-
+#import <Foundation/Foundation.h>
+#import <UIKit/UIKit.h>
 @protocol BranchDeepLinkingControllerCompletionDelegate <NSObject>
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/BranchDeepLinkingController.h]: add a newline before this one?

File: Branch-SDK/Branch-SDK/BranchDeepLinkModel.h [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+//
+//  Created by Parth Kalavadia on 5/15/17.
+//  Copyright © 2017 Parth Kalavadia. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+#import "BranchDeepLinkingController.h"
+@interface BranchDeepLinkModel : NSObject
+
+@property (strong,nonatomic)UIViewController<BranchDeepLinkingController>* viewController;
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/BranchDeepLinkModel.h]: space after comma for argument

File: Branch-SDK/Branch-SDK/Branch.m [hunk 5]
```diff
@@ -1591,29 +1598,68 @@ - (void)handleInitSuccess {
...
+                    
+                    NSAssert([self.deepLinkPresentingController isKindOfClass:[UINavigationController class]]==NO, @"Root View Controller is not of type UINavigationController, hence cannot push");
+                    
+                    break;
+                    
+                default:
+                    if ([self.deepLinkPresentingController isKindOfClass:[UINavigationController class]]) {
+                        if ([self.deepLinkPresentingController respondsToSelector:@selector(showViewController:sender:)]) {
+                            [self.deepLinkPresentingController showViewController:branchSharingController sender:self];
+                        }else{
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/Branch.m]: Can you use spaces as such `} else {` to keep with rest of styling. I see it a couple times throughout the project

Comment (parthkalavadia) [inline Branch-SDK/Branch-SDK/Branch.m]: Yes, I would change it everywhere.

File: Branch-SDK/Branch-SDK/BranchDeepLinkModel.h (hunk 2 above)
Comment (parthkalavadia) [inline Branch-SDK/Branch-SDK/BranchDeepLinkModel.h]: Yes, would surely remove it.

File: Branch-SDK/Branch-SDK/Branch.h (hunk 1 above)
Comment (parthkalavadia) [inline Branch-SDK/Branch-SDK/Branch.h]: Okay, would update to continue supporting to old API.

**Comment by E-B-Smith** [comment]
You can make your updates and resolve the merge conflicts by pulling your branch, editing, merging, and pushing your branch again.

**Comment by parthkalavadia** [comment]
There is a bug for one the use case in the above implementation. I am working on that. Will push it once I have fixed it.

**Comment by BranchMacMini** [comment]
Result of [Integration 4](https://stlt.herokuapp.com/v1/xcs_deeplink/qabot.stage.branch.io/b388335146db7e936b7215574da2f2b2/d85abcdfd92da370bf9c50f785a6a37e)
---
*Duration*: 4 minutes and 3 seconds
*Result*: **Perfect build!** All 179 tests passed. :+1:
*Test Coverage*: 56%

**Comment by BranchMacMini** [comment]
Result of [Integration 5](https://stlt.herokuapp.com/v1/xcs_deeplink/qabot.stage.branch.io/b388335146db7e936b7215574da2f2b2/d85abcdfd92da370bf9c50f785c916f8)
---
*Duration*: 4 minutes and 3 seconds
*Result*: **Perfect build!** All 179 tests passed. :+1:
*Test Coverage*: 56%

**Comment by parthkalavadia** [comment]
I fixed the bug and conflicts. I have even taken care for all the comments by Alex on code review.

File: Branch-SDK/Branch-SDK/Branch.h [hunk 6]
```diff
@@ -409,7 +410,15 @@ typedef NS_ENUM(NSUInteger, BranchCreditHistoryOrder) {
 /// @name Deep Link Controller
 ///---------------------------
 
-- (void)registerDeepLinkController:(UIViewController <BranchDeepLinkingController> *)controller forKey:(NSString *)key;
+- (void)registerDeepLinkController:(UIViewController <BranchDeepLinkingController> *)controller forKey:(NSString *)key __attribute__((deprecated(("This API is deprecated. Please use registerDeepLinkController: forKey: withOption:"))));
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/Branch.h]: Can you link to the documentation about what the new argument is? It's going to be annoying to figure out what the options are without some docs.

File: Branch-SDK/Branch-SDK/Branch.m [hunk 7]
```diff
@@ -1607,29 +1618,110 @@ - (void)handleInitSuccess {
...
-            branchSharingController.deepLinkingCompletionDelegate = self;
-            self.deepLinkPresentingController = [[[UIApplicationClass sharedApplication].delegate window] rootViewController];
             
-            if ([self.deepLinkPresentingController presentedViewController]) {
-                [self.deepLinkPresentingController dismissViewControllerAnimated:NO completion:^{
-                    [self.deepLinkPresentingController presentViewController:branchSharingController animated:YES completion:NULL];
-                }];
+            if([self.deepLinkControllers[key] isKindOfClass:[BranchDeepLinkModel class]]) {
+                BranchDeepLinkModel* deepLinkModel = self.deepLinkControllers[key];
+                UIViewController <BranchDeepLinkingController> *branchSharingController = deepLinkModel.viewController;
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/Branch.m]: This will crash by using the legacy method that you now list as deprecated. Please be more careful about testing these edge cases

Comment (aaustin) [inline Branch-SDK/Branch-SDK/Branch.m]: Oh my mistake. I saw you handle it below

File: Branch-SDK/Branch-SDK/Branch.h (hunk 6 above)
Comment (aaustin) [inline Branch-SDK/Branch-SDK/Branch.h]: Here's the docs that we'd need to update https://dev.branch.io/getting-started/deep-link-routing/advanced/ios/#register-view-controller-for-deep-link-routing

**Comment by BranchMacMini** [comment]
Result of [Integration 6](https://stlt.herokuapp.com/v1/xcs_deeplink/qabot.stage.branch.io/b388335146db7e936b7215574da2f2b2/eec4b3510fa6dc541c675922a302caa4)
---
*Duration*: 3 minutes and 57 seconds
*Result*: **Perfect build!** All 179 tests passed. :+1:
*Test Coverage*: 56%

**Comment by BranchMacMini** [comment]
Result of [Integration 7](https://stlt.herokuapp.com/v1/xcs_deeplink/qabot.stage.branch.io/b388335146db7e936b7215574da2f2b2/eec4b3510fa6dc541c675922a3246bff)
---
*Duration*: 46 seconds
*Result*: **3 errors, failing state: build-errors**

**Comment by BranchMacMini** [comment]
Result of [Integration 8](https://stlt.herokuapp.com/v1/xcs_deeplink/qabot.stage.branch.io/b388335146db7e936b7215574da2f2b2/eec4b3510fa6dc541c675922a3253281)
---
*Duration*: 3 minutes and 39 seconds
*Result*: **Perfect build!** All 179 tests passed. :+1:
*Test Coverage*: 56%

**Comment by aaustin** [comment]
Couple comments still. You view all outstanding comments by scrolling through the "files changed" tab in the PR: https://github.com/BranchMetrics/ios-branch-deep-linking/pull/629/files

**Comment by E-B-Smith** [comment]
Can you update the ChangeLog.md file?  You can add a brief description of your change to the top of the file along with a link to see the full documentation.

**Comment by BranchMacMini** [comment]
Result of [Integration 9](https://stlt.herokuapp.com/v1/xcs_deeplink/qabot.stage.branch.io/b388335146db7e936b7215574da2f2b2/f6b9610889448e5de5fd691f5778d54d)
---
*Duration*: 4 minutes and 10 seconds
*Result*: **Perfect build!** All 179 tests passed. :+1:
*Test Coverage*: 56%

**Comment by BranchMacMini** [comment]
Result of [Integration 10](https://stlt.herokuapp.com/v1/xcs_deeplink/qabot.stage.branch.io/b388335146db7e936b7215574da2f2b2/f6b9610889448e5de5fd691f579897ad)
---
*Duration*: 4 minutes and 16 seconds
*Result*: **Perfect build!** All 179 tests passed. :+1:
*Test Coverage*: 56%

**Comment by BranchMacMini** [comment]
Result of [Integration 11](https://stlt.herokuapp.com/v1/xcs_deeplink/qabot.stage.branch.io/b388335146db7e936b7215574da2f2b2/f6b9610889448e5de5fd691f57bb7fef)
---
*Duration*: 4 minutes and 20 seconds
*Result*: **Perfect build!** All 179 tests passed. :+1:
*Test Coverage*: 56%

File: Branch-SDK/Branch-SDK/Branch.m [hunk 8]
```diff
@@ -1649,18 +1660,114 @@ - (void)handleInitSuccess {
...
+                
+                if ([branchSharingController respondsToSelector:@selector(configureControlWithData:)]) {
+                    [branchSharingController configureControlWithData:latestReferringParams];
+                }
+                else {
+                    BNCLogWarning(@"[Branch Warning] View controller does not implement configureControlWithData:");
```
Comment (E-B-Smith) [inline Branch-SDK/Branch-SDK/Branch.m]: The `BNCLog` functions already format `[Branch Warning]` in a uniform way so they aren't needed in the message.

File: Branch-SDK/Branch-SDK/BranchDeepLinkModel.h [hunk 9]
```diff
@@ -0,0 +1,15 @@
+//
+//  BranchDeepLinkModel.h
+//  Branch-TestBed
+//
+//  Created by Parth Kalavadia on 5/15/17.
+//  Copyright © 2017 Parth Kalavadia. All rights reserved.
+//
+#import <Foundation/Foundation.h>
+#import "BranchDeepLinkingController.h"
+@interface BranchDeepLinkModel : NSObject
```
Comment (E-B-Smith) [inline Branch-SDK/Branch-SDK/BranchDeepLinkModel.h]: Can you call this something more descriptive?  The word 'model' is really overused so it doesn't mean much.

Also, internal classes, classes that the SDK user is unlikely to ever use, start with `BNC` rather than `Branch`.

How about something like `BNCDeepLinkViewControllerInstance`.

File: Branch-SDK/Branch-SDK/BranchDeepLinkingController.h [hunk 10]
```diff
@@ -5,13 +5,21 @@
...
 
-- (void)deepLinkingControllerCompleted;
+- (void)deepLinkingControllerCompleted __attribute__((deprecated(("This API is deprecated. Instead, use deepLinkingControllerCompletedFrom: viewController"))));;
+- (void)deepLinkingControllerCompletedFrom:(UIViewController*) viewController;
 
 @end
 
+typedef NS_ENUM(NSInteger, BNCViewControllerOption) {
```
Comment (E-B-Smith) [inline Branch-SDK/Branch-SDK/BranchDeepLinkingController.h]: How about `BNCViewControllerPresentationOption` ?

File: README.md [hunk 11]
```diff
@@ -385,13 +385,13 @@ Register a controller for Branch to show when specific keys are present in the B
 ###### Objective-C
 
 ```objc
-[[Branch getInstance] registerDeepLinkController:myController forKey:@"my-key"];
+[[Branch getInstance] registerDeepLinkController:myController forKey:@"my-key" withOption:BNCViewControllerOptionShow];
```
Comment (E-B-Smith) [inline README.md]: `withPresentation:`?

File: Branch-TestBed/Branch-TestBed/AppDelegate.m [hunk 12]
```diff
@@ -38,6 +38,13 @@ - (BOOL)application:(UIApplication *)application
     
     [branch setWhiteListedSchemes:@[@"branchtest"]];
 
+    // Automatic Deeplinking on "deeplink_text"
```
Comment (E-B-Smith) [inline Branch-TestBed/Branch-TestBed/AppDelegate.m]: You merged these changes back in on this PR?

**Comment by BranchMacMini** [comment]
Result of [Integration 12](https://stlt.herokuapp.com/v1/xcs_deeplink/qabot.stage.branch.io/b388335146db7e936b7215574da2f2b2/54a4eba29c2f74a2f6b1c16d60b82182)
---
*Duration*: 4 minutes and 29 seconds
*Result*: **Perfect build!** All 179 tests passed. :+1:
*Test Coverage*: 57%

**Comment by BranchMacMini** [comment]
Result of [Integration 13](https://stlt.herokuapp.com/v1/xcs_deeplink/qabot.stage.branch.io/b388335146db7e936b7215574da2f2b2/54a4eba29c2f74a2f6b1c16d60da0bba)
---
*Duration*: 3 minutes and 58 seconds
*Result*: **Perfect build!** All 179 tests passed. :+1:
*Test Coverage*: 57%

**Comment by BranchMacMini** [comment]
Result of [Integration 14](https://stlt.herokuapp.com/v1/xcs_deeplink/qabot.stage.branch.io/b388335146db7e936b7215574da2f2b2/c1fc729633c2dfdd3bd2f3074364530b)
---
*Duration*: 4 minutes and 47 seconds
*Result*: **Perfect build!** All 179 tests passed. :+1:
*Test Coverage*: 57%

**Comment by BranchMacMini** [comment]
Result of [Integration 15](https://stlt.herokuapp.com/v1/xcs_deeplink/qabot.stage.branch.io/b388335146db7e936b7215574da2f2b2/c1fc729633c2dfdd3bd2f307438501d5)
---
*Duration*: 4 minutes and 5 seconds
*Result*: **Perfect build!** All 179 tests passed. :+1:
*Test Coverage*: 57%
